### ISSUE TITLE
[RFC] Save LoRA-only weights during training for inference usage

### ISSUE THREAD

**Issue by guozhihao-224** [verl-project/verl-omni#285]
# RFC: Save LoRA-only weights during training for inference usage

- **Author**: @guozhihao-224
- **Parent roadmap**: verl-project/verl-omni#97 (Q3 Roadmap → Easy-to-Use)
- **Status**: Draft
- **Scope**: `verl_omni` FSDP diffusion path; VeOmni and multi-adapter export are deferred.

## 1. Summary

This RFC adds an adapter-only checkpoint path for LoRA diffusion training. At configured save points, the FSDP diffusion engine gathers the active `default` adapter and publishes a standard PEFT directory:

```text
global_step_<N>/lora_adapter/
  adapter_config.json
  adapter_model.safetensors
  lora_train_meta.json
```

The adapter cadence is independent from full-checkpoint cadence. Full checkpoints, resume behavior, and online rollout weight sync remain unchanged.

1. Reuse `collect_lora_params` as the only FSDP-aware gather path.
2. Produce PEFT-canonical keys in the collector rather than normalizing them after gather.
3. Publish locally through a rank-0 temporary directory and an atomic rename.
4. Treat FSDP collective failures as process-group-fatal, while synchronizing post-gather validation and I/O failures.
5. Require content identity checks before reusing an already-published artifact for the same step.

### Motivation

LoRA training updates only a small set of adapter parameters, but the current trainer checkpoint contains the full actor state, optimizer state, extra state, and dataloader state. This makes frequent checkpoints expensive and leaves inference users with a sharded training artifact instead of a standard adapter directory.

The repository already provides the required building blocks:

- `verl_omni.utils.fsdp_utils.collect_lora_params` gathers LoRA parameters for FSDP1, FSDP2, and non-FSDP modules.
- `DiffusersFSDPEngine.get_per_tensor_param` uses that gather path for online trainer-to-rollout synchronization.
- `NonDiffusersModelBase.load_lora_adapter` and the vLLM-Omni LoRA path consume PEFT-style artifacts.

The missing capability is durable serialization of the gathered adapter.

### Goals

- Save `adapter_config.json` and `adapter_model.safetensors` in standard PEFT format.
- Decouple adapter save frequency from full-checkpoint frequency.
- Preserve existing full-checkpoint semantics for all users.
- Keep FSDP summon, offload, and rank discipline inside the engine.
- Provide an independent adapter discovery pointer without changing training resume semantics.

### Non-Goals

- VeOmni support before its engine supports LoRA training.
- Multi-adapter export; phase 1 exports only `default`.
- Changes to online rollout IPC or `TensorLoRARequest` / `OmniTensorLoRARequest`.
- LoRA merge tooling.
- Remote or HDFS adapter publication. Phase 1 is local-only because the current storage abstraction has no atomic remote directory commit primitive.

## 3. Current Gap

| Capability | Current state |
|---|---|
| LoRA configuration and lifecycle | Exists in `DiffusionModelConfig` and `LoRAAdapterMixin` |
| FSDP-aware LoRA gather | Exists in `collect_lora_params` |
| Online trainer-to-rollout sync | Exists and is not changed by this RFC |
| Full checkpoint save | Exists through `ray_diffusion_trainer._save_checkpoint` |
| PEFT adapter load | Exists in `NonDiffusersModelBase` and vLLM-Omni integration |
| Adapter-only disk serialization | Missing |

The current inference loader is intentionally defensive: it intersects saved keys with the model's expected PEFT keys and warns instead of raising on mismatches. The exporter must therefore guarantee key alignment; otherwise an adapter can load successfully while silently retaining initialized weights.

### 4.1 Configuration and save triggers

Introduce a diffusion-specific `DiffusionCheckpointConfig(CheckpointConfig)` with:

```yaml
actor_rollout_ref:
  actor:
    checkpoint:
      _target_: verl_omni.workers.config.diffusion.DiffusionCheckpointConfig
      save_contents: ['model', 'optimizer', 'extra']
      save_lora_adapter: false
      save_lora_adapter_freq: 0
```

`save_lora_adapter_freq` is validated as non-negative:

- `0` follows the full-checkpoint trigger.
- A positive value creates an independent adapter cadence.
- A negative value is rejected during configuration validation.
- `is_last_step` always triggers an enabled adapter save.
- `save_freq == -1` still disables full checkpoints. With a positive adapter frequency, adapter-only training remains valid.

The trainer should centralize the two predicates instead of duplicating them in each fit loop:

```text
trigger_full = save_freq > 0 and (
    is_last_step or step % save_freq == 0 or esi_close_to_expiration
)

trigger_lora = save_lora_adapter and has_lora and (
    is_last_step
    or esi_close_to_expiration
    or (save_lora_adapter_freq > 0 and step % save_lora_adapter_freq == 0)
    or (save_lora_adapter_freq == 0 and trigger_full)
)
```

`lora_adapter_path` with `lora_rank <= 0` remains invalid for the current FSDP engine, because the engine only enters LoRA construction when `lora_rank > 0`.

### 4.2 Adapter export contract

Add `DiffusersFSDPEngine.save_lora_checkpoint(local_path, global_step=0, layered_summon=False)`. The engine writes directly into the complete `local_path` supplied by the trainer; it does not append another `lora_adapter` component.

The method follows this contract:

1. Resolve the active `default` PEFT config.
2. Perform a collective presence check before any rank returns. If any rank lacks the config, every rank returns `skipped` and no rank enters the FSDP gather.
3. Load the model to GPU when required by the existing offload policy.
4. Call `collect_lora_params(..., base_sync_done=True, is_diffusers=True, adapter_name="default", key_format="peft")`.
5. Cast tensors for export only after gather. `lora_dtype=None` preserves the current dtype; configured values use `PrecisionType.to_dtype`.
6. Validate keys, serialize config and metadata, and exchange validation results with `all_gather_object`.
7. Write the artifact on rank 0, broadcast the result, execute the final barrier, and restore offload state in an unconditional cleanup path.

#### PEFT configuration

`adapter_config.json` starts from the active `peft_config.to_dict()` and must preserve fields required by both standard PEFT loaders and the installed PEFT version, including:

- `peft_type`;
- `base_model_name_or_path`, `revision`, `task_type`, and `inference_mode` when present;
- LoRA fields such as `r`, `lora_alpha`, `target_modules`, `target_parameters`, `exclude_modules`, and `bias` when configured and supported.

The serializer must detect fields that are rejected or silently dropped by the installed PEFT version. If `target_parameters` or `exclude_modules` cannot be represented without changing the expected key set, export fails before a directory is published. verl-omni-only values such as `lora_dtype`, `fsdp_layer_prefixes`, and `policy_state_adapters` belong in `lora_train_meta.json`, not in `adapter_config.json`.

The format test must exercise both `LoraConfig.from_dict` and the standard `PeftConfig` / `PeftModel.from_pretrained` path.

#### Canonical keys

`collect_lora_params` gains `key_format="rollout" | "peft"`, defaulting to the current rollout behavior. The PEFT path owns key construction because it knows the FSDP version, summoned submodule, and block prefix.

- Non-layered FSDP1 and non-FSDP output uses PEFT keys after removing only known wrapper artifacts.
- Layered FSDP1 combines the exact block prefix with the PEFT-relative key once.
- FSDP2 follows the same rule at the submodule boundary and does not duplicate an already-present prefix.
- The exporter does not perform generic `.default.` replacement.
- Keys containing `_fsdp_wrapped_module` or `.base_layer`, empty collections, and duplicate canonical keys are errors.
- Duplicate detection happens at insertion time, before an `OrderedDict` update can overwrite an earlier value. Errors include both source paths and the canonical key.

### 4.3 Worker and trainer integration

Expose the engine method through both worker layers:

```python
class TrainingWorker(Worker):
    @register(dispatch_mode=Dispatch.ONE_TO_ALL)
    def save_lora_checkpoint(self, local_path, global_step=0, layered_summon=False):
        return self.engine.save_lora_checkpoint(
            local_path, global_step, layered_summon=layered_summon
        )

class ActorRolloutRefWorker(Worker):
    @register(dispatch_mode=Dispatch.ONE_TO_ALL)
    def save_lora_checkpoint(self, local_path, global_step=0, layered_summon=False):
        assert "actor" in self.role
        return self.actor.save_lora_checkpoint(
            local_path, global_step, layered_summon=layered_summon
        )
```

The trainer passes `actor_rollout_ref.rollout.layered_summon` explicitly in both standalone and colocated modes. It must not depend on `self.layered_summon`, which is initialized only when the worker also owns a rollout.

`_save_lora_checkpoint` writes only the adapter artifact. It does not write `data.pt`, update `latest_checkpointed_iteration.txt`, or prune adapter directories. The full-save block runs first; the adapter-save block runs afterward. The trainer updates `latest_lora_adapter_iteration.txt` only after worker status succeeds and driver-side artifact validation passes.

### 4.4 Artifact layout and consumers

```text
<default_local_dir>/
  global_step_<N>/
    actor/
    lora_adapter/
      adapter_config.json
      adapter_model.safetensors
      lora_train_meta.json
  latest_checkpointed_iteration.txt
  latest_lora_adapter_iteration.txt
```

`default_local_dir` must be visible to actor rank 0 and the trainer driver. Non-local paths and deployments explicitly marked node-local are invalid when adapter saving is enabled. A normal local path is still checked from the driver before the discovery pointer is advanced.

The artifact is intended for:

- `NonDiffusersModelBase.load_lora_adapter`;
- `PeftModel.from_pretrained` where the base module supports the PEFT API;
- vLLM-Omni offline LoRA loading;
- validated diffusers model and pipeline loader APIs.

## 5. Failure, Atomicity, and Artifact Identity

The save protocol has two failure classes.

### Collective failures

Exceptions inside `collect_lora_params` or other FSDP collectives are process-group-fatal. They propagate through the WorkerGroup RPC and abort training; they must not be converted into a rank-local status.

### Post-gather failures

After gather, every rank follows the same sequence:

1. Validate keys, dtype, and serialized metadata locally.
2. Compute an `artifact_schema_fingerprint` from keys, dtype/shape metadata, and canonical JSON.
3. Compute an `artifact_content_fingerprint` from sorted keys, dtype/shape, tensor payload bytes, and canonical adapter config content. The fingerprint excludes its own metadata field.
4. Store the content fingerprint in `lora_train_meta.json` and exchange validation results with `all_gather_object`.
5. Fail validation if any rank failed or successful ranks disagree on schema or content.
6. If validation succeeds, rank 0 writes to a unique sibling temporary directory, fsyncs files and directories, and renames into an absent destination.
7. Broadcast the rank-0 write result, execute the final barrier, and always restore offload state and clear caches.

The published directory is immutable. A retry for the same step is idempotent only when the existing directory is complete and its stored content fingerprint matches the current artifact. An incomplete, step-mismatched, or content-mismatched directory is never overwritten. `validate_complete_adapter` recomputes the published fingerprint before the driver writes `latest_lora_adapter_iteration.txt`.

## 6. Alternatives Considered

| Alternative | Decision | Reason |
|---|---|---|
| Add a `'lora'` content type to upstream `CheckpointConfig.save_contents` | Defer | Requires an upstream change and couples adapter cadence to full-checkpoint cadence. |
| Serialize entirely in the trainer | Reject | Duplicates FSDP summon, offload, and rank discipline outside the engine. |
| Store under `actor/lora_adapter/` | Reject | Couples adapter lifecycle to full-checkpoint retention and makes adapter-only shipping less direct. |
| Extend upstream `_save_lora_train_meta` | Defer | Keeps the wrong ownership and still couples the save to full-checkpoint cadence. |
| Call `get_peft_model_state_dict` again on the wrapped module | Reject | Risks a second FSDP traversal; the collector already has the required provenance. |

### Format and unit tests

- Non-FSDP round-trip through `NonDiffusersModelBase.load_lora_adapter` with bit-exact weights and no missing/unexpected key warnings.
- `LoraConfig.from_dict`, `PeftConfig.from_pretrained`, and `PeftModel.from_pretrained` all accept the artifact.
- Configured `target_parameters` and `exclude_modules` either survive the PEFT round-trip with matching keys or fail before publication.
- FSDP1 full summon, FSDP1 layered summon, and FSDP2 key sets match an equivalent non-FSDP reference.
- Standalone and colocated worker RPCs both pass `layered_summon=True` to the engine.
- `key_format="rollout"` remains compatible with existing online sync tests; `key_format="peft"` has no wrapper artifacts, empty keys, or silent collisions.
- `lora_dtype="fp32"` exports fp32 tensors; `lora_dtype=None` preserves dtype.
- A rank-local missing PEFT config produces synchronized `skipped` results without entering FSDP gather.

### Integration and failure tests

- With adapter frequency 2 and full frequency 10, step 2 has only `lora_adapter/`, while step 10 has both artifacts.
- Adapter-only steps do not write `data.pt`; full-checkpoint and adapter pointers advance independently.
- Full-checkpoint retention removes old `actor/` paths without removing sibling adapters.
- Inject rank-0 write errors, one-rank validation errors, FSDP collective errors, and driver-invisible paths; verify the documented fail-closed behavior.
- Retry a complete same-step artifact with matching content fingerprint and repair only the pointer; reject a different fingerprint without overwrite.
- Validate PEFT, `NonDiffusersModelBase`, vLLM-Omni, and selected diffusers pipeline loading.

### Rollout

- **Phase 1**: FSDP diffusion, one `default` adapter, synchronous local publication.
- **Phase 2**: VeOmni after LoRA training support lands.
- **Phase 3**: Multi-adapter export with a manifest.
- **Phase 4**: Remote publication after the storage layer gains a commit primitive.
- **Phase 5**: Upstream `verl` AR FSDP integration.

### Compatibility

- New fields default to `false` and `0`; existing users see no behavior change.
- Existing `actor/` checkpoint layout and resume pointer remain unchanged.
- Adapter-only artifacts do not participate in training resume.
- Upstream `actor/lora_train_meta.json` remains for backward compatibility.
- `async_save` does not make diffusion FSDP saving asynchronous in this phase.

### Future work

- Multi-adapter export and adapter-specific pruning.
- Adapter merge CLI.
- Diffusion-specific inference metadata.
- Remote adapter discovery and publication.

**Comment by wtomin** [comment]
I think this LoRA-only weight saving support is crucial. However, I am not sure if it is necessary to save LoRA-only weight during training, which can be more time consuming. As far as I know, `verl` only convert fsdp checkpoints to PEFT LoRA weights after training is finished.

As for the APIs, I suggest to keep consistent with `verl`.  `verl.model_merge` is the entrypoint for converting fsdp checkpoints to PEFT LoRA weights. Maybe we can inherit from `verl.model_merge.FSDPModelMerger` and support diffusion models' weights conversion.

**Comment by NancyFyong** [comment]
Hi @guozhihao-224 ，Thank you very much for your contribution and outstanding work. Is there any further update?
